### ISSUE TITLE
DataLoader.shuffle() should reshuffle data at each epoch

### ISSUE THREAD

**Issue by L-M-Sherlock** [burn-rs/burn#701]
### Feature description

DataLoader.shuffle() should reshuffle data at every epoch at each epoch.

### Feature motivation

By shuffling the data at each epoch, the model is exposed to different patterns in each epoch, which helps prevent it from overfitting to specific patterns in the training,

### (Optional) Suggest a Solution

In PyTorch, it's implemented by random sampler: https://github.com/pytorch/pytorch/blob/0b868b19063645afed59d6d49aff1e43d1665b88/torch/utils/data/sampler.py#L40-L82
